### ISSUE TITLE
Constructor(Type.ctor) type problems

### ISSUE THREAD

**Issue by iDevelopThings** [Hookyns/ts-reflection#12]
Not sure if this is a known issue or not, was going to figure a way to fix it if so.
Also saw some TODO in the code, so I'm not sure if you have some issues in mind etc.

This is my TS:
```typescript
import {SomeServiceClass} from "./tests/Classes/ControllerSetup/SomeServiceClass";
import {getType} from "tst-reflect";

const someServiceClass = getType<SomeServiceClass>();

const ctor = someServiceClass.ctor;

const t = Reflect.construct(ctor, []);
```

This is the generated JS:
```javascript
"use strict";
Object.defineProperty(exports, "__esModule", {value : true});
const tst_reflect_1 = require("tst-reflect");
(0, tst_reflect_1.getType)({
	k     : 0,
	n     : "SomeServiceInterface",
	fn    : "@envuso/container/src/tests/Classes/ControllerSetup/SomeServiceInterface.ts:SomeServiceInterface#21493",
	props : [{n : "propertyName", t : (0, tst_reflect_1.getType)({n : "string", k : 2}), am : 2, acs : 0, ro : false, o : false}],
	meths : [{n : "randomMethod", params : [], rt : (0, tst_reflect_1.getType)({n : "any", k : 2}), tp : [], o : false, am : 2}],
}, 21493);
(0, tst_reflect_1.getType)({
	k     : 1,
	n     : "SomeServiceClass",
	fn    : "@envuso/container/src/tests/Classes/ControllerSetup/SomeServiceClass.ts:SomeServiceClass#21497",
	props : [{n : "propertyName", t : (0, tst_reflect_1.getType)({n : "string", k : 2}), am : 2, acs : 0, ro : false, o : false}],
	meths : [{
		n      : "getMessage",
		params : [],
		rt     : (0, tst_reflect_1.getType)({n : "string", k : 2}),
		tp     : [],
		o      : false,
		am     : 2,
	}, {n : "randomMethod", params : [], rt : (0, tst_reflect_1.getType)({n : "void", k : 2}), tp : [], o : false, am : 2}],
	ctors : [{params : []}],
	ctor  : () => SomeServiceClass_1.SomeServiceClass,
	iface : (0, tst_reflect_1.getType)(21493),
}, 21497);
const someServiceClass = (0, tst_reflect_1.getType)(21497);
const ctor             = someServiceClass.ctor;
const t                = Reflect.construct(ctor, []);
//# sourceMappingURL=Test.js.map

```

Essentially the import for "SomeServiceClass" is removed from the file 

So when we call Reflect.construct(ctor, []) it cannot reference the actual module and this error is thrown:
```ReferenceError: SomeServiceClass_1 is not defined```

When I use the metadata file, it just becomes null instead 🤔

**Comment by iDevelopThings** [comment]
As a side note/idea

I was thinking it may be useful to add some other meta related to file location, I have some code in the project I was originally building which output, the source file location & where the built file is

This way you could dynamically Import this file, which would potentially be useful for the `Type.ctor` since you could just output something like

```typescript
	/**
	 * Constructor function in case Type is class
	 */
	get ctor(): Function | undefined {
		return () => require(this._fileLocation)[this._name];
	}
```

If you're open to the idea, I'll submit a pr, it's pretty simple :D

**Comment by iDevelopThings** [comment]
> Essentially the import for "SomeServiceClass" is removed from the file
> 
> So when we call Reflect.construct(ctor, []) it cannot reference the actual module and this error is thrown: `ReferenceError: SomeServiceClass_1 is not defined`

Was just playing some more... and this might actually just be the typescript compiler that does this, because "SomeServiceClass" isn't actually used, just used as a type

If i add ```const inst = new SomeServiceClass()``` to the ts file it works fine

**Comment by Hookyns** [comment]
This is odd. I did not know TS itself do optimizations like this..

You cannot do just dynamic import, because there can be not exported classes. 
But this is relared
https://github.com/Hookyns/ts-reflection/blob/27da5872bb34d92987cff1422fdba31365e4a075/transformer/src/getTypeDescription.ts#L292-L301

**Comment by iDevelopThings** [comment]
> This is odd. I did not know TS itself do optimizations like this..
> 
> You cannot do just dynamic import, because there can be not exported classes. But this is relared
> 
> https://github.com/Hookyns/ts-reflection/blob/27da5872bb34d92987cff1422fdba31365e4a075/transformer/src/getTypeDescription.ts#L292-L301

Yeah me neither, this is a new one to me 😅

And yeah, I have a PR in the working for this, hopefully it solves the problem and you agree with my solution 👍

**Comment by iDevelopThings** [comment]
@Hookyns take a look at that commit and let me know what you think

https://github.com/iDevelopThings/ts-reflection/commit/f4ebbeccf51a82b364ed2663f01053c43f6c40fb

I didn't go too overkill but it works quite nicely from what i can tell so far, the only problem i foresee is if people don't configure the default tsconfig "outdir" setting or have it slightly different than i ever expected

**Comment by Hookyns** [comment]
@iDevelopThings I've just realized.. this will break front-end usage, won't it? Webpack will not process this without special plugin, right? We should rather generate direct imports into transpiled .js source files.

**Comment by iDevelopThings** [comment]
> this will break front-end usage, won't it? Webpack will not process this without special plugin, right? 

F I didn't think of this 🤦 

I was trying to do that originally, it wouldn't work for some reason though, maybe I was modifying the usage of the metafile though, not 100% sure 😅

**Comment by iDevelopThings** [comment]
Ahh... its not far off working, just a couple of things to work out :D

![image](https://user-images.githubusercontent.com/4105581/151557132-35b19d69-8b11-4c8a-a822-f403e950d964.png)

**Comment by iDevelopThings** [comment]
I wonder if this would work with webpack? 🤔

![image](https://user-images.githubusercontent.com/4105581/151562209-9942571e-fc29-4936-95f4-bc68a5c82015.png)

For some reason I cant get the top-level require name to match/reference the name in `ctor`

Even if i pass the identifier down into the node creation for `ctor` which is how I thought it worked for a kind of "unique" identifier 😅

**Comment by Hookyns** [comment]
I think this should work.

**Comment by iDevelopThings** [comment]
Okay sweet - atleast from a container perspective this would give a way to resolve those instances 😅

Are you cool with keeping the `ctorDesc` meta for other purposes? As you say, it's not so useful from a browser perspective, but for node servers, it can be so handy 

There's probably some stuff I can remove from it also so it's less to output

**Comment by Hookyns** [comment]
It increase the size of metadata but it's OK. Maybe I would reduce those properties just to `en` (exportName) and `path` with absolute path. No relativePath nor requirePath... relative to what? and require path can be absolute path.

It's just odd, why is `default` name of that type (I mean `n: "default",`? It is name of the export but name of the type is name of that class. It must be bug in my original code..

**Comment by iDevelopThings** [comment]
Yeah, I wonder if it would be better to have another function similar to getType() for getting that module location information instead, so it's only included when requested... but eh

relativePath is the path relative to where the meta is, for example, src/Nested/Deeper/ClassA.ts and src/ClassB.ts
In ClassA, it would be "./../../ClassB.ts" - basically how you would define the path for an import yourself. But it's not necessary now with this latest change 😄 

And yeah i noticed that with the name there, we also don't need this export anymore

Maybe... just this data? naming can be changed up as you wish

```javascript {
	srcPath : "/Users/sam/Code/Packages/ts-reflection/dev/testing/src/ctor-reflection/Nested/NestedClass.ts",
	outPath : "/Users/sam/Code/Packages/ts-reflection/dev/testing/dist/ctor-reflection/Nested/NestedClass.js",
	en      : "default",
	n       : "NestedClass",
}```

srcPath/outPath just makes life easier when you're working with the files in all kinds of situations, outPath could be used to do a dynamic require like my original code on a node server

**Comment by Hookyns** [comment]
Well, lets introduce new config property, `mode: "universal" | "server" = "universal"`.
Universal is basic usage for both server and client. But there will be no redundant information. Server will be extended.
Does it sound good?

**Comment by iDevelopThings** [comment]
That makes sense to me 😄 good idea^^

I also wonder, are you open to using an outDir in the tsconfig of both packages?

These js files here can be annoying at times, especially when you need to delete them all 😅
![image](https://user-images.githubusercontent.com/4105581/151573829-651b884e-0cfb-48f3-9e9f-0f54bbe81631.png)

**Comment by Hookyns** [comment]
> I also wonder, are you open to using an outDir in the tsconfig of both packages?

NP, change it.

**Comment by Hookyns** [comment]
Merged in #18
